text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1848 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Introduce new marker traits, `MetaSized` and `Pointee`, adding it to a trait
+hierarchy with the [`Sized`][api_sized] trait, and make `Sized` and `MetaSized`
+traits `const`:
+
+```
+       ┌────────────────┐                          ┌─────────────────────────────┐
+       │ const Sized    │ ───────────────────────→ │ Sized                       │
+       │ {type, target} │         implies          │ {type, target, runtime env} │
+       └────────────────┘                          └─────────────────────────────┘
+               │                                                  │
+            implies                                            implies
+               │                                                  │
+               ↓                                                  ↓
+┌──────────────────────────────┐             ┌───────────────────────────────────────────┐
+│ const MetaSized              │ ──────────→ │ MetaSized                                 │
+│ {type, target, ptr metadata} │   implies   │ {type, target, ptr metadata, runtime env} │
+└──────────────────────────────┘             └───────────────────────────────────────────┘
+                                                                  │
+                                                               implies
+                                                                  │
+                                      ┌───────────────────────────┘
+                                      ↓
+                            ┌──────────────────┐
+                            │ Pointee          │
+                            │ {runtime env, *} │
+                            └──────────────────┘
+```
+
+Or, in Rust syntax:
+
+```rust
+#![feature(const_trait_impl)]
+
+#[const_trait] trait Sized: ~const MetaSized {}
+
+#[const_trait] trait MetaSized: Pointee {}
+
+trait Pointee {}
+```
+
+## Implementing `Sized`
+[implementing-sized]: #implementing-sized
+
+Implementations of the proposed traits are automatically generated by
+the compiler and cannot be implemented manually:
+
+- `Pointee`
+    - Types that which can be used from behind a pointer (they may or may
+      not have a size).
+    - `Pointee` will be implemented for:
+        - `MetaSized` and `const MetaSized` types
+        - `extern type`s from [rfcs#1861][rfc_extern_types]
+        - compound types where every element is `Pointee`
+    - In practice, every type will implement `Pointee`.
+- `MetaSized`
+    - Types whose size is computable given pointer metadata, and knowledge of
+      the type, target platform and runtime environment.
+      - `const MetaSized` does not require knowledge of the runtime
+        environment
+    - `MetaSized` is a subtrait of `Pointee`
+    - `MetaSized` will be implemented for:
+        - `Sized` types
+        - slices `[T]` where every element is `MetaSized`
```

Comment:
```suggestion
        - slices `[T]` where every element is `Sized`
```

Looks like a typo...

Reply:
Fixed in 7a11fe901b0d977e26ee7749b230f05c01ba53bd